kernel/src/sched.rs

```diff
@@ -332,69 +436,117 @@ impl Kernel {
+        &self,
         platform: &P,
         chip: &C,
         ipc: Option<&ipc::IPC>,
+        scheduler: &SC,
         _capability: &dyn capabilities::MainLoopCapability,
-    ) {
+    ) -> ! {
         chip.watchdog().setup();
         loop {
             chip.watchdog().tickle();
             unsafe {
-                chip.service_pending_interrupts();
-                DynamicDeferredCall::call_global_instance_while(|| !chip.has_pending_interrupts());
-
-                for p in self.processes.iter() {
-                    chip.watchdog().tickle();
-                    p.map(|process| {
-                        self.do_process(platform, chip, process, ipc);
-                    });
-                    if chip.has_pending_interrupts()
-                        || DynamicDeferredCall::global_instance_calls_pending().unwrap_or(false)
-                    {
-                        break;
+                match scheduler.kernel_work_ready(chip) {
+                    true => {
+                        // Execute kernel work
+                        scheduler.execute_kernel_work(chip);
                     }
-                }
-
-                chip.atomic(|| {
-                    if !chip.has_pending_interrupts()
-                        && !DynamicDeferredCall::global_instance_calls_pending().unwrap_or(false)
-                        && self.processes_blocked()
-                    {
-                        chip.watchdog().suspend();
-                        chip.sleep();
-                        chip.watchdog().resume();
+                    false => {
+                        // No kernel work ready, so ask scheduler for a process
+                        match scheduler.next(self) {
+                            SchedulingDecision::RunProcess((appid, timeslice_us)) => {
+                                self.process_map_or((), appid, |process| {
+                                    let (reason, time_executed) = self.do_process(
+                                        platform,
+                                        chip,
+                                        scheduler,
+                                        process,
+                                        ipc,
+                                        timeslice_us,
+                                    );
+                                    scheduler.result(reason, time_executed);
+                                });
+                            }
+                            SchedulingDecision::Sleep => {
+                                chip.atomic(|| {
+                                    // cannot sleep if interrupts are pending, as on most platforms
+                                    // unhandled interrupts will wake the device.
+                                    if !chip.has_pending_interrupts()
+                                        && !DynamicDeferredCall::global_instance_calls_pending()
+                                            .unwrap_or(false)
```

Comment:
Isn't the other case that the chip never wakes up again?

Reply:
I think the other case is perhaps that the chip doesn't wake until a _different_ interrupt triggers (e.g. the scheduler chooses to ignore a peripheral, go to sleep, but wants to wake for a timer)

Reply:
I don't think we could just remove the check, because of the scenario where an interrupt arrives while `scheduler.next()` is executing and returning `Sleep`. We could add *another* scheduler trait function with a default impl there, if we want to support sleeping despite outstanding (pending but disabled) interrupts. I don't feel super strongly either way

Reply:
I propose that if a use case comes around that demands that level of control, we could revisit this, but I think it is good to have some protective defaults at this low layer.